Debugging in docker

rharris-nitro opened the pull request:
Sets up debugging environments for main production app and testing
- creates debugger (debugpy) in .vscode/launch.json and Taskfile
- divvies up Dockerfile into 2 images main (prod) and dev
- also adds docker-compose implementation

rlasskiy commented:
it looks like something wrong with this PR

follow committed is from different PR

[changes <> to []](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/724414dee944e1c6285387d6737c48b8a9d5f4b5)
[created Taskfile and added build, run, and test commands](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/ca59d1bc3d4143a72f8b6d739a9d362ce12ef1c5)
[creates taskfile setup documentation](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/81349c2d94e5cf15e4f8c24e8a3a3e72ef669205)
[moves the task: No tasks with description available. Try --list-all t…](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/f05ecd43f2eb17682c2ca403a89de79afc6e601d)
[includes building and running with task](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/ad9ece0d2e14ce8f07e379676f52c68a77650a3e)
[reformatting headers](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/f461b00f2970db1dd8e0e6d0fbfaf61fd42e6582)
[reformatting headers](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/0ad22d6b2e1aef06245279ff7f858a9cf6bb356d)
[adds global variables to task](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/bd2b74465759390173e48ad51d5d84ea41e44f58)
[gets rid of docker commands section as task has made it easier](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/c49e69d8f793326bbbf660aecacf2ce8482f7e9f)
[gets rid of task setup page](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/f3ab62082b69058285bf8bf25e8c1b966408c9d6)

rharris-nitro commented:
> it looks like something wrong with this PR
> 
> follow committed is from different PR
> 
> [changes <> to []](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/724414dee944e1c6285387d6737c48b8a9d5f4b5) [created Taskfile and added build, run, and test commands](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/ca59d1bc3d4143a72f8b6d739a9d362ce12ef1c5) [creates taskfile setup documentation](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/81349c2d94e5cf15e4f8c24e8a3a3e72ef669205) [moves the task: No tasks with description available. Try --list-all t…](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/f05ecd43f2eb17682c2ca403a89de79afc6e601d) [includes building and running with task](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/ad9ece0d2e14ce8f07e379676f52c68a77650a3e) [reformatting headers](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/f461b00f2970db1dd8e0e6d0fbfaf61fd42e6582) [reformatting headers](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/0ad22d6b2e1aef06245279ff7f858a9cf6bb356d) [adds global variables to task](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/bd2b74465759390173e48ad51d5d84ea41e44f58) [gets rid of docker commands section as task has made it easier](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/c49e69d8f793326bbbf660aecacf2ce8482f7e9f) [gets rid of task setup page](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/f3ab62082b69058285bf8bf25e8c1b966408c9d6)

fixed this issue by deleting commits using `git rebase -i`

rlasskiy commented on requirements.txt:
```diff
@@ -0,0 +1,4 @@
+# To ensure app dependencies are ported from your virtual environment/host machine into your container, run 'pip freeze > requirements.txt' in the terminal to overwrite this file
+
+debugpy==1.8.0
```
![image](https://github.com/rharris-nitro/pythonExercise1/assets/16574236/e5a2c4d4-3a9e-43e9-845e-565511fd2c16)

rlasskiy commented on .dockerignore:
```diff
@@ -0,0 +1,28 @@
+**/secrets.dev.yaml
+**/values.dev.yaml
+LICENSE
+README.md
```
??

rlasskiy commented on Dockerfile:
```diff
@@ -4,17 +4,24 @@ FROM python:3.11 AS base
+COPY app/ .
+
+ENTRYPOINT ["python"]
+CMD ["app/main.py"]
```
CMD ["main.py"]

rlasskiy commented on docker-compose.test.yml:
```diff
@@ -0,0 +1,8 @@
+    build:
+      context: .
+      target: test
+    image: my-python-test
```
please add possibility to debug the test as well

rlasskiy commented on Taskfile.yml:
```diff
@@ -1,22 +1,20 @@
-
 tasks:
-  build-app:
+  build:
```
documentation is reference 
  ```bash
  task build-app
  ```

rlasskiy commented on .gitignore:
```diff
@@ -1,2 +1,15 @@
+.vscode/*
+!.vscode/settings.json
+!.vscode/tasks.json
+!.vscode/launch.json
```
`.vscode/launch.json` is missing in this PR

rlasskiy commented on Dockerfile:
```diff
@@ -4,17 +4,24 @@ FROM python:3.11 AS base
 
-    CMD ["pytest", "test_main.py"]
+ENTRYPOINT ["pytest"]
+CMD ["test_main.py"]
```
do you need this?
Do you plan to extend this each time and then add a new test file?

rlasskiy commented on docker-compose.yml:
```diff
@@ -0,0 +1,11 @@
+    image: my-python-app
+    ports:
+      - "5673:5673"
+    command: ${BASE_COMMAND:- main.py}
```
during debugging, we debug files which uploaded to the container during `docker build` process
we need to add the possibility to debug files from the working directory

rlasskiy commented on .vscode/launch.json:
```diff
@@ -0,0 +1,14 @@
+      ]
+    }
+  ]
+}
```
you add [.vscode/launch.json](https://github.com/rharris-nitro/pythonExercise1/pull/9/files#diff-bd5430ee7c51dc892a67b3f2829d1f5b6d223f0fd48b82322cfd45baf9f5e945) but your documentation `Setup local environment` not mention what we need install VSCode

rlasskiy commented on Taskfile.yml:
```diff
@@ -1,22 +1,20 @@
+      - docker-compose -f docker-compose.test.yml build
 
-  run-test:
+  up-test:
```
you change the name of the task but not update the documentation which is still reference
  ```bash
  task run-test
  ```

rlasskiy commented on docs/local-environment.md:
````diff
@@ -97,3 +87,15 @@ test_main.py ..                                                          [100%]
+task up USE_DEBUG=true
+```
+
+The same can be done for the test image.
````
Attach to debugger via VSCode step is missing

rlasskiy commented on docs/local-environment.md:
```diff
@@ -50,37 +51,29 @@ brew install go-task/tap/go-task
+### Install VSCode<a name="install-vscode"></a>
 
-### Run App<a name="run-app"></a>
+You should open the folder in VSCode in the directory where the repo is cloned.
```
Please add the link to VSCode

rlasskiy commented on docs/debugging.md:
```diff
@@ -0,0 +1,39 @@
+
+## VSCode Configuration:
+
+In VSCode, you need to configure a launch configuration that tells VSCode how to connect to the debugger inside the Docker container. This involves specifying the type of debugger (e.g., Python), the host (usually localhost or 127.0.0.1), and the port number that matches the one configured in the Docker container.
```
add link to `.vscode/launch.json`
